File: demo4/src/main/java/examples/p01/SimpleThread.java
package examples.p01;

import lombok.*;

public class SimpleThread {
    @SneakyThrows
    public static void main(String[] args) throws InterruptedException {
        String message = "Hello KSUG";
        ThreadSample t1 = new ThreadSample(message);

        // main thread가 중요한 이유
//        t1.setDaemon(true);
        t1.start();
//        t1.join();

        Thread t2 = new Thread(new Runnable() {
            @Override
            @SneakyThrows
            public void run() {
                Thread.sleep(100);
                System.out.println("==============================");
                System.out.println("message from Runnable: " + message);
                System.out.println("==============================");

                Thread t = Thread.currentThread();
                long id = t.getId();
                String name = t.getName();
                int priority = t.getPriority();
                Thread.State state = t.getState();
                ThreadGroup threadGroup = t.getThreadGroup();
                String groupName = threadGroup.getName();
                ThreadGroup parent = t.getThreadGroup().getParent();
                String parentName = parent.getName();

                System.out.println("id: " + id);
                System.out.println("namd :" + name);
                System.out.println("priority: " + priority);
                System.out.println("state: " + state);
                System.out.println("threadGroup.name : " + groupName);
                System.out.println("parentName : " + parentName);

                Thread.sleep(1000);
            }
        });
//        t2.setDaemon(true);

        t2.start();
//        t2.join();

        System.out.println("exit main thread");
    }

}

class ThreadSample extends Thread {

    private String message;

    public ThreadSample(String message) {
        this.message = message;
    }

    @Override
    @SneakyThrows
    public void run() {
        Thread.sleep(500);
        System.out.println("==============================");
        System.out.println("message from ThreadSample: " + message);
        System.out.println("==============================");


        Thread t = Thread.currentThread();
        long id = t.getId();
        String name = t.getName();
        int priority = t.getPriority();
        State state = t.getState();
        ThreadGroup threadGroup = t.getThreadGroup();
        String groupName = threadGroup.getName();
        ThreadGroup parent = t.getThreadGroup().getParent();
        String parentName = parent.getName();

        System.out.println("id: " + id);
        System.out.println("namd :" + name);
        System.out.println("priority: " + priority);
        System.out.println("state: " + state);
        System.out.println("threadGroup.name : " + groupName);
        System.out.println("parentName : " + parentName);

        Thread.sleep(1000);
    }
}

File: kotlin/src/study/thread/SimpleThread.kt
package study.thread.p01

fun main(args: Array<String>) {
    var message = "Hello KSUG"
    var tread1 = ThreadSample(message)
    tread1.setDaemon(true)
    tread1.start()
    tread1.join()

    var runnable = Runnable {
        Thread.sleep(500)
        println("message from ThreadSample: ${message}")
        showThreadInfo()
        Thread.sleep(500)
    }

    var thread2 = Thread( runnable )
    thread2.setDaemon(true)
    thread2.start()
    thread2.join()

}

fun showThreadInfo() {
        val t = Thread.currentThread()
        val text = """
            |==============================
            |thread
            | - id: ${t.id}
            | - name: ${t.name}
            | - priority: ${t.priority}
            | - state: ${t.state}
            | - groupName: ${t.threadGroup.name}
            | - parentName: ${t.threadGroup.parent.name}
            |==============================
        """.trimMargin()
        println(text)
}

class ThreadSample(message: String): Thread() {
    val message = message

    override fun run() {
        Thread.sleep(500)
        println("message from ThreadSample: ${message}")
        showThreadInfo()
        Thread.sleep(500)
    }
}